### PR TITLE
Support for the new pandas rolling API.

### PR DESCRIPTION
**Pull request by mikegraham**
Should support basically the same subset of rolling as the old dd.rolling_* functions.

@jcrist

### CODE REVIEW

File: dask/dataframe/tests/test_rolling.py [hunk 1]
```diff
@@ -36,6 +39,27 @@ def rolling_tests(p, d):
...
+    eq(p.rolling(3).max(), d.rolling(3).max())
+    eq(p.rolling(3).std(), d.rolling(3).std())
+    eq(p.rolling(3).var(), d.rolling(3).var())
+    eq(p.rolling(3).skew(), d.rolling(3).skew())
+    eq(p.rolling(3).kurt(), d.rolling(3).kurt())
+    eq(p.rolling(3).quantile(0.5), d.rolling(3).quantile(0.5))
+    eq(p.rolling(3).apply(mad), d.rolling(3).apply(mad))
+    with ignoring(ImportError):
+        eq(p.rolling(3, win_type='boxcar').sum(),
+           d.rolling(3, win_type='boxcar').sum())
```
Comment (mrocklin) [inline dask/dataframe/tests/test_rolling.py]: Why was the ImportError here ignored originally?  If it is for older pandas versions then we probably don't need to worry about this, given the dependence on recent pandas for this feature. 

Conversely, we may expect this test to fail under the older Pandas version tests (do we still want to support these?)

File: dask/dataframe/tests/test_rolling.py [hunk 2]
```diff
@@ -36,6 +39,27 @@ def rolling_tests(p, d):
     # Test with kwargs
     eq(pd.rolling_sum(p, 3, min_periods=3), dd.rolling_sum(d, 3, min_periods=3))
 
+    # New rolling API
```
Comment (mrocklin) [inline dask/dataframe/tests/test_rolling.py]: We might want this to be a separate test

File: dask/dataframe/tests/test_rolling.py [hunk 3]
```diff
@@ -36,6 +39,27 @@ def rolling_tests(p, d):
...
+    eq(p.rolling(3).quantile(0.5), d.rolling(3).quantile(0.5))
+    eq(p.rolling(3).apply(mad), d.rolling(3).apply(mad))
+    with ignoring(ImportError):
+        eq(p.rolling(3, win_type='boxcar').sum(),
+           d.rolling(3, win_type='boxcar').sum())
+    # Test with edge-case window sizes
+    eq(p.rolling(0).sum(), d.rolling(0).sum())
+    eq(p.rolling(1).sum(), d.rolling(1).sum())
+    # Test with kwargs
+    eq(p.rolling(3, min_periods=2).sum(), d.rolling(3, min_periods=2).sum())
```
Comment (mrocklin) [inline dask/dataframe/tests/test_rolling.py]: These tests look great by the way.

File: dask/dataframe/rolling.py [hunk 4]
```diff
@@ -57,3 +57,78 @@ def rolling(arg, window, *args, **kwargs):
...
+    else:
+        method = getattr(this_partition.rolling(window), method_name)
+        return method(*method_args, **method_kwargs)
+
+def wrap_rolling_method(name):
+    def inner(self, *args, **kwargs):
+        return self._call_method(name, *args, **kwargs)
+    return inner
+
+class Rolling(object):
```
Comment (mrocklin) [inline dask/dataframe/rolling.py]: This class probably needs a docstring.  Some options:

1.  Write a custom docstring explaining how it works, this is good if there are significant differences between this and Pandas
2.  Use `@wraps(pd....Rolling)` to steal the Pandas docstring

File: dask/dataframe/rolling.py [hunk 5]
```diff
@@ -57,3 +57,78 @@ def rolling(arg, window, *args, **kwargs):
...
+        applied = method(*method_args, **method_kwargs)
+        return applied.iloc[extra:]
+    else:
+        method = getattr(this_partition.rolling(window), method_name)
+        return method(*method_args, **method_kwargs)
+
+def wrap_rolling_method(name):
+    def inner(self, *args, **kwargs):
+        return self._call_method(name, *args, **kwargs)
+    return inner
```
Comment (mrocklin) [inline dask/dataframe/rolling.py]: I'm curious to see how well this will work when serialized.  Dynamically generated functions are sometimes odd.

One option would be to use `toolz.curry` to build these from an `inner` function that was in the main namespace.  Also, the inner function should probably have a better name.  It will be used within diagnostics.

```python
def rolling_part(...):  # previously called inner
    ...

sum = curry(rolling_part, 'sum')
```

This is just one option though, and probably not the best one.

File: dask/dataframe/rolling.py [hunk 6]
```diff
@@ -57,3 +57,78 @@ def rolling(arg, window, *args, **kwargs):
...
+def call_pandas_rolling_method_single(this_partition, rolling_kwargs,
+        method_name, method_args, method_kwargs):
+    # used for the start of the df/series
+    if this_partition.shape[0] < rolling_kwargs['window']:
+        raise NotImplementedError("Window larger than partition size")
+
+    method = getattr(this_partition.rolling(**rolling_kwargs), method_name)
+    return method(*method_args, **method_kwargs)
+
+def call_pandas_rolling_method_with_neighbor(prev_partition, this_partition,
```
Comment (mrocklin) [inline dask/dataframe/rolling.py]: These function names are long.  These names get used in user diagnostics (like the multi-progress bars) so it's nice to have them be a concise if possible.

File: dask/dataframe/core.py [hunk 7]
```diff
@@ -1480,6 +1502,10 @@ def groupby(self, key, **kwargs):
         from dask.dataframe.groupby import DataFrameGroupBy
         return DataFrameGroupBy(self, key, **kwargs)
 
+    @derived_from(pd.DataFrame)
+    def rolling(self, window, min_periods=None, win_type=None):
+        return super(DataFrame, self).rolling(window, min_periods, win_type)
```
Comment (mrocklin) [inline dask/dataframe/core.py]: Ah, we're re-implementing these here just for the different docstring, yes?

File: dask/dataframe/core.py [hunk 8]
```diff
@@ -577,6 +577,24 @@ def _aca_agg(self, token, func, aggfunc=None):
         """ Wrapper for aggregations """
         raise NotImplementedError
 
+    def rolling(self, window, min_periods=None, win_type=None):
+        from dask.dataframe.rolling import Rolling
+
+        if not isinstance(window, int):
+            raise TypeError('window must be an integer')
+        if window < 0:
+            raise ValueError('window must be positive')
```
Comment (mrocklin) [inline dask/dataframe/core.py]: Should we test these?

```python
with pytest.raises(ValueError):
    df.rolling(-1).max()
```

File: dask/dataframe/rolling.py (hunk 6 above)
Comment (mrocklin) [inline dask/dataframe/rolling.py]: My mistake, this isn't true, the `rolling-` prefix is what will be used

File: dask/dataframe/core.py (hunk 7 above)
Comment (mikegraham) [inline dask/dataframe/core.py]: Yes.

Comment (mrocklin) [inline dask/dataframe/core.py]: I'm not suggesting anything here.  Just building intuition.

**Comment by mrocklin** [comment]
Test failures are just for older versions of Pandas.  We should probably skip the tests if the Pandas version is < 0.18.0.  This can be done with `@pytest.mark.skipif(condition, reason="some text")`.

File: dask/dataframe/tests/test_rolling.py (hunk 2 above)
Comment (mikegraham) [inline dask/dataframe/tests/test_rolling.py]: Updated, separating out `dd.rolling_foo(...)` tests from `ddf.rolling(...).foo(...)` tests.

File: dask/dataframe/rolling.py (hunk 4 above)
Comment (mikegraham) [inline dask/dataframe/rolling.py]: I didn't add one yet. You were going to reflect/research what this should look like, let me know when you've got a chance. @jreback related that there's no real reason for someone to call `Rolling` themselves, so I'm still not sure that `Rolling` is the best place to document things rather than `Series.rolling`/`DataFrame.rolling`. The docstring for pandas' `Rolling` class is mostly about the `__init__` I don't currently mimic -- I can do so in case someone goes searching around if that's what you think is best.

File: dask/dataframe/rolling.py (hunk 5 above)
Comment (mikegraham) [inline dask/dataframe/rolling.py]: I pushed a commit that changed these to static code.

File: dask/dataframe/core.py (hunk 7 above)
Comment (mikegraham) [inline dask/dataframe/core.py]: I ended up putting a reasonable docstring in `_Frame`.

File: dask/dataframe/core.py (hunk 8 above)
Comment (mikegraham) [inline dask/dataframe/core.py]: I added such tests.

File: dask/dataframe/rolling.py [hunk 9]
```diff
@@ -120,15 +118,38 @@ def _call_method(self, method_name, *args, **kwargs):
...
-    median = wrap_rolling_method('median')
-    min = wrap_rolling_method('min')
-    max = wrap_rolling_method('max')
-    std = wrap_rolling_method('std')
-    var = wrap_rolling_method('var')
-    skew = wrap_rolling_method('skew')
-    kurt = wrap_rolling_method('kurt')
-    quantile = wrap_rolling_method('quantile')
-    apply = wrap_rolling_method('apply')
+    def count(self, *args, **kwargs):
```
Comment (mikegraham) [inline dask/dataframe/rolling.py]: I should probably give these buggers their pandas docstrings too.

File: dask/dataframe/tests/test_rolling.py [hunk 10]
```diff
@@ -61,20 +62,33 @@ def rolling_tests(p, d):
     # Test with kwargs
     eq(p.rolling(3, min_periods=2).sum(), d.rolling(3, min_periods=2).sum())
 
+def test_rolling_functions_series():
```
Comment (mrocklin) [inline dask/dataframe/tests/test_rolling.py]: You'll probably have to add the following decorator

```python
@pytest.mark.skipif(LooseVersion(pd.__version__) <= '0.18.0',
                    reason="rolling object not supported")
```

See `dask/dataframe/tests/test_dataframe.py` for examples.

**Comment by mrocklin** [comment]
You can safely ignore the Python 2.7 errors.  These are due to #1227

File: dask/dataframe/tests/test_rolling.py (hunk 10 above)
Comment (mikegraham) [inline dask/dataframe/tests/test_rolling.py]: I added the appropriate skips, I think, updated the PR.

File: dask/dataframe/rolling.py (hunk 6 above)
Comment (jcrist) [inline dask/dataframe/rolling.py]: Why the need for two methods here (the second being a superset of the first)? In the previous version we got by with just a single `rolling_chunk` method.

File: dask/dataframe/core.py [hunk 11]
```diff
@@ -577,6 +577,40 @@ def _aca_agg(self, token, func, aggfunc=None):
...
+
+        if min_periods is not None:
+            if not isinstance(min_periods, int):
+                raise TypeError('min_periods must be an integer')
+            if min_periods < 0:
+                raise ValueError('min_periods must be positive')
+
+        return Rolling(self, {'window': window,
```
Comment (jcrist) [inline dask/dataframe/core.py]: This is an api change without a deprecation of the old api (may break user code). For the other pandas api change such as `resample` we implemented deprecation logic based on pandas version (see https://github.com/dask/dask/blob/master/dask/dataframe/tseries/resample.py#L19) - perhaps the same should be done here?

Comment (jcrist) [inline dask/dataframe/core.py]: The old function api could also be changed to rely on the new `Rolling` object.

Comment (mikegraham) [inline dask/dataframe/core.py]: I left the old code untouched, figuring that would guarantee 0 change in behavior and would let the pandas deprecation warning bubble up, but it's obviously easy enough to do so. I don't know dask's policies with respect to deprecated features in pandas.

Comment (jcrist) [inline dask/dataframe/core.py]: Oop, I feel silly. Completely forgot that the old version didn't have `rolling` as a method on dataframe. And seeing as the `rolling_*` functions are deprecated in pandas, leaving those alone is probably fine as well. Sorry for the noise here, this is fine as is.

**Comment by mrocklin** [comment]
This all looks pretty good to me.  I'm +1 to merge.  @jcrist any concerns?

Something to think about for the future would be to explicitly split out the overlapping part of the previous partition as a separate task.  This would complicate the graph a bit but would also help to reduce communication significantly, making this more appropriate for the distributed memory case.

File: dask/dataframe/core.py [hunk 12]
```diff
@@ -577,6 +577,40 @@ def _aca_agg(self, token, func, aggfunc=None):
...
+            Minimum number of observations in window required to have
+            a value (otherwise result is NA).
+        win_type: string, default None
+            Provide a window type. (Identical to pandas.)
+
+        The center, freq, and axis arguments are not supported.
+        """
+        from dask.dataframe.rolling import Rolling
+
+        if not isinstance(window, int):
```
Comment (jreback) [inline dask/dataframe/core.py]: pandas already does these kinds of validations

what is the purpose of duplicating here?

**Comment by mikegraham** [comment]
A couple potentially outstanding things

- the timing and exception types and wording for invalid input aren't the same as pandas, I can update those (though pandas gets the timing wrong on one of them probably)
- there's still an exception case I don't test, I can fix that
- you thought I maybe should have documented the Rolling class itself
- I didn't update the actual Sphinx documentation, I can do so if you want.

**Comment by jcrist** [comment]
>  @jcrist any concerns?

Only thing is the comment above about two functions where one used to suffice. There's some duplication there between the two that could be removed, but it's not a big deal. Merging without fixing that is fine.

File: dask/dataframe/rolling.py [hunk 13]
```diff
@@ -57,3 +57,99 @@ def rolling(arg, window, *args, **kwargs):
...
+        for i in range(1, self.obj.npartitions + 1):
+            dsk[new_name, i] = (
+                call_pandas_rolling_method_with_neighbor,
+                (old_name, i-1), (old_name, i),
+                self.rolling_kwargs, method_name, args, kwargs)
+
+        return self.obj._constructor(
+            merge(self.obj.dask, dsk),
+            new_name,
+            self.obj,
```
Comment (mrocklin) [inline dask/dataframe/rolling.py]: OK, so this argument defines the metadata of the output.  It should be an empty dataframe (or series) with the dtypes and column names that we expect the output to have.  If everything is the same then we can just pass the input in directly, however if things are likely to change (such as if we take the rolling mean of an integer-valued series then this should change a bit.

Typically this is done by calling the Pandas version of the function on `self.obj._pd` which is an empty Pandas object that mimics the shape of the output.  So it might be wise to replace this with something like the following:

    metadata = self.obj._pd.rolling(window=1, ...).method_name()

If there is an issue with the Pandas object being empty then we might want to steal @necaris 's solution in https://github.com/dask/dask/pull/1222

File: dask/dataframe/core.py (hunk 12 above)
Comment (mrocklin) [inline dask/dataframe/core.py]: It's nice to raise an exception immediately at graph construction if we can rather than wait for the user to execute the graph later.

**Comment by mrocklin** [comment]
All of those things sound nice to have.  They could happen now or later as you prefer.  Adding to sphinx would be highly appreciated.

File: dask/dataframe/core.py (hunk 12 above)
Comment (jreback) [inline dask/dataframe/core.py]: his would raise as soon as u construct the .rolling object on the pandas side - it's lazy as well

Comment (mrocklin) [inline dask/dataframe/core.py]: Right. The problem is that we only construct the `.rolling` object during computation.  This error might happen quite a bit later and quite possibly on another machine.  It's nice to raise the error when the user enters the code rather than possibly quite a bit later.

Comment (jreback) [inline dask/dataframe/core.py]: I understand; I would immediately construct it and then throw it away (so if it doesn't raise then u r good)

otherwise when say rolling starts allows strings (soon) you will have to implement the same logic as pandas has duplicating efforts

Comment (mrocklin) [inline dask/dataframe/core.py]: Ah I see.  Yes, I agree that would be good.  Perhaps [this comment](https://github.com/dask/dask/pull/1221/files/a740f8155ebd0e5b1214daf75eca7fbfd4bdd0da#r65988750) is a good place to do this?

File: dask/dataframe/rolling.py (hunk 13 above)
Comment (mikegraham) [inline dask/dataframe/rolling.py]: Thanks, I'd seen the `_pd` attribute and not known what it was about. I just played monkey-see-monkey-do with constructing the graph there.

Oddly, even adding integer columns the tests still pass, but I can look at updating this appropriately.

File: dask/dataframe/core.py (hunk 12 above)
Comment (mikegraham) [inline dask/dataframe/core.py]: Oddly enough, pandas doesn't raise for everything when you call `rolling`, but when you call an actual method sometimes

```
In [138]: df.rolling(-1)
Out[138]: Rolling [window=-1,center=False,axis=0]
```

Nonetheless, I can probably just call the method on the `_pd` too to get the exceptions to bubble..?

Comment (mikegraham) [inline dask/dataframe/core.py]: (Actually, that wouldn't work, it doesn't raise on an empty.)

Comment (jreback) [inline dask/dataframe/core.py]: can u raise an issue for the negative integers on pandas tracker - that should immediately raise

**Comment by mikegraham** [comment]
I pushed one more thing incorporating a bit more of what I discussed. I ended up not updating the documentation; I had made some changes but ultimately was not certain what the right thing to do was because I don't understand dask's attitude toward pandas version support.

**Comment by mrocklin** [comment]
Merged

### Future work

* https://github.com/dask/dask/issues/1243
* https://github.com/dask/dask/issues/1242
